src/cmd-cloud-prune

```diff
@@ -0,0 +1,291 @@
+# pruning actions required for each stream and the age threshold for
+# deleting artifacts within them.
+# Example of policy.yaml
+# rawhide:
+#     # all cloud images
+#     cloud-uploads: 2 years
+#     # artifacts in meta.json's `images` key
+#     images: 2 years
+#     images-keep: [qemu, live-iso]
+#     build: 3 years
+# The script also updates the builds.json for the respective stream by
+# adding the policy-cleanup key when we set the upload_builds_json flag.
+# It adds the relevant actions completed to that key
+# For eg:
+#     "builds": [
+#         {
+#             "id": "40.20240425.dev.1",
+#             "arches": [
+#                 "x86_64"
+#             ],
+#             "policy-cleanup": [
+#                 "cloud-uploads",
+#                 "images-kept": ["qemu", "live-iso"]
+#             ]
+#         }
+
+import argparse
+import json
+import re
+import pytz
+import yaml
+import collections
+import datetime
+import os
+import boto3
+from dateutil.relativedelta import relativedelta
+from cosalib.gcp import remove_gcp_image
+from cosalib.aws import deregister_aws_resource
+from cosalib.builds import BUILDFILES
+from cosalib.s3 import s3_copy
+
+Build = collections.namedtuple("Build", ["id", "images", "arch"])
+# set metadata caching to 5m
+CACHE_MAX_AGE_METADATA = 60 * 5
+
+
+def main():
+    # Parse arguments and initialize variables
+    args = parse_args()
+    with open(BUILDFILES['sourceurl'], "r") as file:
+        builds_source_data_url = file.read()
+    bucket, prefix = parse_s3_url(builds_source_data_url)
+    cloud_config = get_cloud_config(args)
+    stream = args.stream
+
+    # Boto3 loads credentials from ~/.aws/config by default and we can change
+    # this default location by setting the AWS_CONFIG_FILE environment variable.
+    # The Python bindings don't support passing a config file.
+    # The alternative is to manually pass ACCESS_KEY and SECRET_KEY which isn't favourable.
+    if args.aws_config_file:
+        os.environ["AWS_CONFIG_FILE"] = args.aws_config_file
+    s3_client = boto3.client("s3")
+
+    # These lists are up to date as of schema hash
+    # 4c19aed3b3d84af278780bff63728510bb3e70613e4c4eef8cabd7939eb31bd8. If changing
+    # this hash, ensure that the list of supported and unsupported artifacts below
+    # is up to date.
+    supported = ["amis", "gcp"]
+    unsupported = ["aliyun", "azurestack", "digitalocean", "exoscale", "ibmcloud", "powervs", "azure"]
+
+    with open(args.policy, "r") as f:
+        policy = yaml.safe_load(f)
+    validate_policy(stream, policy)
+
+    with open(BUILDFILES['list'], "r") as f:
+        builds_json_data = json.load(f)
+
+    # Prune builds based on the policy
+    for action in ['cloud-uploads', 'images', 'build']:
+        if action not in policy[stream]:
+            continue
+        duration = get_period_in_months(policy[stream][action])
+        ref_date = datetime.datetime.now() - relativedelta(months=int(duration))
```

Comment:
Minor: let's call `.now()` once outside the loop instead of multiple times. That will ensure we're consistently comparing against the same date across iterations.